libslide/src/math/gcd_poly_zz.rs

```diff
@@ -0,0 +1,430 @@
+    if res_f.is_empty() {
+        return poly_negate(&res_g);
+    }
+    if res_g.is_empty() {
+        return res_f;
+    }
+    let df = res_f.len() as isize - 1;
+    let dg = res_g.len() as isize - 1;
+    let mut h: Poly;
+    let k: usize;
+    if df == dg {
+        for i in (0..res_f.len()).rev() {
+            res_f[i] -= res_g[i];
+        }
+        truncate_zeros(&res_f)
+    } else {
+        k = ((df - dg) as f64).abs() as usize;
+        if df > dg {
+            h = res_f[..k].to_vec();
+            res_f = res_f[k..].to_vec();
+        } else {
+            h = poly_negate(&res_g[..k].to_vec());
+            res_g = res_g[k..].to_vec();
+        }
+        for i in 0..min(f.len(), g.len()) {
+            res_f[i] -= res_g[i];
+        }
+        h.append(&mut res_f);
+        // truncate zeros
+        truncate_zeros(&h)
+    }
+}
+
+#[allow(unused)]
+fn truncate_zeros(f: &PolyPtrArg) -> Poly {
+    let mut res = f.to_vec();
+    let mut count = 0;
+    for i in &res {
+        if *i == 0 {
+            count += 1;
+        } else {
+            break;
+        }
+    }
+    res = res.into_iter().rev().collect();
+    res.truncate(res.len() - count);
+    res.into_iter().rev().collect()
+}
+
+#[allow(unused)]
+fn poly_div(f: &PolyPtrArg, g: &PolyPtrArg) -> (Poly, Poly) {
+    let res_f = f.to_vec();
+    let res_g = g.to_vec();
+    let df = res_f.len() as isize - 1;
+    let dg = res_g.len() as isize - 1;
+    if res_g.is_empty() {
+        panic!("Zero Division Error in Pol_Div");
+    } else if df < dg {
+        return ([].to_vec(), res_f);
+    }
+    let lc_g = lc(&res_g);
+    let mut rem = res_f;
+    let mut dr = df;
+    let mut dr2: isize;
+    let mut c_pow: isize;
+    let mut lc_r: isize;
+    let mut quo: Poly = [].to_vec();
+    let mut heu: Poly;
+    loop {
+        lc_r = lc(&rem);
+        // currently this only supports integer division
+        // todo Expand this to handle fractions
+        if lc_r % lc_g != 0 {
+            break;
+        }
+        if (lc_r as f64 / lc_g as f64) - (lc_r / lc_g) as f64 != 0.0 {
+            panic!("Exact division failed")
+        } else {
+            c_pow = lc_r / lc_g;
+        }
+        quo = poly_add_term(&quo, c_pow, dr - dg);
+        heu = poly_mul_term(&res_g, c_pow, dr - dg);
+        rem = poly_sub(&rem, &heu);
+
+        dr2 = dr;
+        dr = rem.len() as isize - 1;
+        if dr < dg {
+            break;
+        } else if dr >= dr2 {
+            panic!("Polynomial Division Failed");
+        }
+    }
+    (quo, rem)
+}
+
+#[cfg(test)]
+mod test {
+    use super::gcd_poly_zz_heu;
+    use super::poly_add_term;
+    use super::poly_div;
+    use super::poly_mul_term;
+    use super::poly_sub;
+    use super::prim_poly;
+
+    #[test]
+    fn test_gcd_poly_zz_heu() {
+        assert_eq!(
+            gcd_poly_zz_heu(vec![1, 0, -1], vec![1, -3, 2]),
+            Some((vec![1, -1], vec![1, 1], vec![1, -2]))
+        )
+    }
```

Comment:
After this PR is ready I'm gonna come in and hijack the test format.. if that's okay with you ;)